packages/flutter/lib/src/services/text_input.dart

```diff
@@ -856,25 +856,51 @@ abstract class TextInputClient {
   ///
   /// [TextInputClient] should cleanup its connection and finalize editing.
   void connectionClosed();
+
+  /// Framework notified of a text input [source] that has been updated.
+  ///
+  /// [TextInputClient] should re-attach itself if it wants to show the text
+  /// input control from the new text input source.
+  ///
+  /// See also:
+  ///
+  ///  * [TextInput.attach], a method used to re-attach the client.
+  ///  * [TextInput.setSource], a method used to change the text input source.
+  void didUpdateInputSource(TextInputSource source);
 }
 
 /// An interface for interacting with a text input control.
 ///
+/// [TextInputConnection] communicates with the platform text input plugin
+/// over the [SystemChannels.textInput] method channel. See [SystemChannels.textInput]
+/// for more details about the method channel messages.
+///
+/// ### Connect to a different input source
+///
+/// The default [TextInputConnection] can be replaced by setting a custom
```

Comment:
nit: this sounds slightly inaccurate, `setSource` doesn't "replace" the connection, but the developer may decide to replace the connection in `didUpdateInputSource`, is that correct?